File: sdks/java/endurance-tests/src/test/java/inc/reactor/sdk/endurance/AudioPublishSteadyTest.java
package inc.reactor.sdk.endurance;

import inc.reactor.sdk.Reactor;
import inc.reactor.sdk.Track;
import java.time.Duration;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

/**
 * The same shape as video publish steady, for audio.
 *
 * <p>A separate scenario rather than one parameterised over both: they share nothing below
 * {@code pushFrame} — separate adapters, separate encoder threads in the native runtime — so a leak
 * in one is not evidence about the other, and a report naming {@code audio-publish-steady} is worth
 * more than one saying {@code media-publish-steady[audio]}.
 */
final class AudioPublishSteadyTest {

    /** 20 ms at 48 kHz mono: the frame size every Reactor model takes. */
    private static final short[] SILENCE = new short[48_000 / 50];

    @Test
    @DisplayName("audio publish steady")
    void audioPublishSteady() throws Exception {
        String apiKey = Endurance.apiKey();
        Endurance.run(
                "audio-publish-steady",
                "publish once and hold it, pushing PCM for the whole run — no pause, no unpublish, " + "no reconnect.",
                false,
                run -> {
                    Reactor reactor = Endurance.connected(apiKey);
                    try {
                        Track microphone = reactor.track("mic");
                        microphone.publish().join();

                        while (run.keepGoing()) {
                            for (int index = 0; index < 50 && run.keepGoing(); index++) {
                                microphone.pushFrame(SILENCE, 48_000, 1);
                                Endurance.pause(Duration.ofMillis(20));
                            }
                            if (!microphone.isPublished()) {
                                throw new AssertionError("the publish did not survive the run");
                            }
                            run.endOfCycle();
                        }
                        microphone.unpublish();
                    } finally {
                        try {
                            reactor.disconnect().join();
                        } catch (RuntimeException alreadyGone) {
                            // The report still has to be written.
                        }
                        reactor.close();
                    }
                });
    }
}

File: sdks/android/endurance-tests/src/androidTest/kotlin/inc/reactor/sdk/android/endurance/AudioPublishSteadyTest.kt
package inc.reactor.sdk.android.endurance

import androidx.test.ext.junit.runners.AndroidJUnit4
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import org.junit.Test
import org.junit.runner.RunWith
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * The same shape as video publish steady, on the audio path.
 *
 * Kept separate rather than parameterised: video and audio share nothing below `pushFrame` — a
 * different native entry point, a different encoder, a different buffer discipline — so a leak in
 * one would be invisible in the other's run.
 */
@RunWith(AndroidJUnit4::class)
internal class AudioPublishSteadyTest {
    private companion object {
        const val SAMPLE_RATE = 48_000
        const val CHANNELS = 1

        /** 10ms per push, the frame size WebRTC's pipeline is built around. */
        const val SAMPLES_PER_CHANNEL = 480
        const val FRAME_INTERVAL_MS = 10L
    }

    @Test
    fun audioPublishSteady() =
        runBlocking {
            val apiKey = Endurance.apiKey()
            Endurance.run(
                "audio-publish-steady",
                "publish once and hold it, pushing 10ms PCM buffers for the whole run — no pause, " +
                    "no unpublish, no reconnect.",
                liveClientsBaselineZero = false,
            ) { run ->
                val reactor = Endurance.connected(apiKey)
                try {
                    val input = reactor.track("mic")
                    input.publish()
                    val pcm =
                        ByteBuffer
                            .allocateDirect(SAMPLES_PER_CHANNEL * CHANNELS * 2)
                            .order(ByteOrder.nativeOrder())

                    while (run.keepGoing()) {
                        pcm.rewind()
                        input.pushFrame(pcm, SAMPLES_PER_CHANNEL, SAMPLE_RATE, CHANNELS)
                        delay(FRAME_INTERVAL_MS)

                        if (!input.published) {
                            throw AssertionError("the held track stopped being published mid-run")
                        }
                        run.endOfCycle()
                    }
                } finally {
                    runCatching { reactor.disconnect() }
                    reactor.close()
                }
            }
        }
}
